### PR TITLE
replace relations with maps

### LINKED ISSUE #462: replace relations with forward maps
Replace relations with forward maps. Follow-up to https://github.com/opendp/opendp/discussions/386.

We would like to also use forward maps in measurements.
The current proposal for maps for multi-parameter privacy distances is to pass forward a privacy curve:
```rust
impl<Q: Clone> Measure for SmoothedMaxDivergence<Q> {
    type Distance = SMDCurve<Q>;
}

struct SMDCurve<Q> {
    epsilon: Rc<dyn Fn(&Q) -> Fallible<Q>>,
    delta: Rc<dyn Fn(&Q) -> Fallible<Q>>,
}

impl<Q: Clone> Measure for FixedSmoothedMaxDivergence<Q> {
    type Distance = (Q, Q);
}
```

I think it may make more sense to use the following approach instead:
```rust
struct Smoothed<Q, M: Measure> { dist_type: PhantomData::<Q>, meas_type: PhantomData::<M> }
impl<Q, M: Measure> Measure for Smoothed<Q, M> {
    type Distance = Rc<dyn Fn(&Q) -> Fallible<M::Distance>>;
}

struct FixedSmoothed<Q, M: Measure> { dist_type: PhantomData::<Q>, meas_type: PhantomData::<M> }
impl<Q, M: Measure> Measure for FixedSmoothed<Q, M> {
    type Distance = (Q, M::Distance);
}
```

This way any existing privacy measure can be augmented to include a delta term for failure events.

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #462

* The analytic gaussian mechanism was particularly tricky to port. I ended up working out an adjusted algorithm that solves for epsilon, instead of sigma.
* Added a `make_fix_delta` combinator
* Implemented TotalOrd for AnyObject to preserve check functionality
    * TotalOrd for AnyObject doesn't use lexicographic ordering, which was previously a footgun in check

### CODE REVIEW

**Review by andrewvyrros — COMMENTED** [review]
This is a big one, glad we can finally make this transition! It's great to see the simplifications in `core.rs`. A lot of surrounding code was touched, of necessity with a change like this, but it's surprisingly manageable. The fact that we'd made heavy use of `new_from_constant()` for relations made the switch trivial in a lot of cases. Having a new incarnation of `check()` is a good idea for similar reasons.

I still need to work through some of the new measurement implementations, but I've been through all the core code, and wanted to submit the review up to this point.

File: rust/src/data/bootstrap.json [hunk 1]
```diff
@@ -85,5 +85,17 @@
             }
         ],
         "ret": {"c_type": "FfiResult<void *>"}
+    },
+    "smd_curve_epsilon": {
+        "description": "Internal function. Use an SMDCurve to find epsilon at a given `delta`.",
+        "args": [
+            {"name": "curve", "c_type": "const AnyObject *"},
+            {"name": "delta", "c_type": "const AnyObject *", "rust_type": "T"},
+            {"name": "T", "is_type": true}
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/data/bootstrap.json:94]: It's a bug in the grammar rules: https://github.com/github/linguist/issues/5943

File: rust/src/core/ffi.rs [hunk 2]
```diff
@@ -6,8 +6,8 @@ use std::os::raw::c_char;
 use crate::{try_, try_as_ref};
 use crate::error::{Error, ErrorVariant, ExplainUnwrap, Fallible};
 use crate::ffi::any::{AnyMeasurement, AnyObject, AnyTransformation, IntoAnyMeasurementExt, IntoAnyTransformationExt};
-use crate::ffi::util;
-use crate::ffi::util::{c_bool, into_c_char_p};
+use crate::ffi::util::{self, c_bool};
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/core/ffi.rs:9]: magical!

File: rust/src/comb/chain/ffi.rs [hunk 3]
```diff
@@ -50,7 +50,7 @@ mod tests {
             Function::new(|arg: &T| arg.clone()),
             SymmetricDistance::default(),
             MaxDivergence::default(),
-            PrivacyRelation::new(|_d_in, _d_out| true),
+            PrivacyMap::new(|_d_in| f64::INFINITY),
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/comb/chain/ffi.rs:53]: Can I make this change in the sequential composition PR? I merged these test measurement constructors there.

File: rust/src/data/ffi.rs [hunk 4]
```diff
@@ -203,7 +205,8 @@ pub extern "C" fn opendp_data__object_as_slice(obj: *const AnyObject) -> FfiResu
                 dispatch!(hashmap_to_raw, [(K, @hashable), (V, @primitives)], (obj))
             } else { fallible!(FFI, "unrecognized generic {:?}", name) }
         }
-        _ => { dispatch!(plain_to_raw, [(&obj.type_, @primitives)], (obj)) }
+        // This list is explicit because it allows us to avoid including u32 in the @primitives
+        _ => { dispatch!(plain_to_raw, [(&obj.type_, [u8, u32, u64, u128, i8, i16, i32, i64, i128, usize, f32, f64, bool])], (obj)) }
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/data/ffi.rs:209]: We do indeed. Nice catch.

**Review by andrewvyrros — COMMENTED** [review]
I've reviewed all the measurements, they look good. The one remaining thing is analytic gaussian; I've sanity checked that, and it seems solid, but it'd be good to have another set of eyes check the math.

**Review by andrewvyrros — APPROVED** [review]
Really excited for this change, it makes a lot of things cleaner.

**Pull request merged**